lib/clAdapters/office365Adapter.js

```diff
@@ -0,0 +1,303 @@
+    headers : {
+      Authorization: 'Bearer ' + token,
+      Accept: 'application/json;odata.metadata=none'
+    }
+  };
+
+  return rp(emailRequestOptions)
+  .then(function(body) {
+    result.success = true;
+    result.data = JSON.parse(body);
+    return true;
+  })
+  .catch(function(err) {
+    result.success = false;
+    result.errorMessage = JSON.stringify(err);
+    return true;
+  });
+};
+
+var getAccessToken = function(clientId, tenantId, certThumbprint, privateKey) {
+  var tokenRequestUrl = 'https://login.microsoftonline.com/' + tenantId + '/oauth2/token?api-version=1.0';
+
+  var jwtHeader = {
+    'alg': 'RS256',
+    'x5t': certThumbprint
+  };
+
+  var jwtPayload = {
+    'aud': tokenRequestUrl,
+    'exp': ((new Date()).getTime() / 1000) + 360000, // add X hours
+    'iss': clientId,
+    'jti': guid.create(),
+    'nbf': ((new Date()).getTime() / 1000) - 360000, // add -X hours
+    'sub': clientId
+  };
+
+  var encodedJwtHeader = new Buffer(JSON.stringify(jwtHeader)).toString('base64');
+  var encodedJwtPayload = new Buffer(JSON.stringify(jwtPayload)).toString('base64');
+  var stringToSign = encodedJwtHeader + '.' + encodedJwtPayload;
+
+  //sign it!
+  var signer = crypto.createSign('RSA-SHA256');
+  signer.update(stringToSign);
+  var encodedSignedJwtInfo = signer.sign(privateKey, 'base64');
+
+  //define assertion
+  var clientAssertion = encodedJwtHeader + '.' + encodedJwtPayload + '.' + encodedSignedJwtInfo;
+
+  var tokenRequestFormData = {
+    client_id: clientId,
+    client_assertion_type: 'urn:ietf:params:oauth:client-assertion-type:jwt-bearer',
+    grant_type: 'client_credentials',
+    resource: 'https://outlook.office365.com/',
+    client_assertion: clientAssertion
+  };
+
+  var tokenRequestOptions = {
+    method: 'POST',
+    port: 443,
+    uri: 'https://login.microsoftonline.com/' + tenantId + '/oauth2/token?api-version=1.0',
+    formData: tokenRequestFormData,
+  };
+
+  return rp(tokenRequestOptions)
+  .then(function(body) {
+    var tokenData = JSON.parse(body);
+    if (tokenData && tokenData.access_token) {
+      var accessToken = tokenData.access_token;
+      return accessToken;
+    } else {
+      return Q.reject('Could not get access token.');
+    }
+  })
+  .catch(function(err) {
+    var tokenData = JSON.parse(JSON.stringify(err));
+    if (tokenData.name === 'StatusCodeError') {
+      var entireMessage = tokenData.message;
+      var messageJson = entireMessage.replace(tokenData.statusCode + ' - ', '');
+      var messageData = JSON.parse(messageJson.replace(new RegExp('\\"', 'g'),'"'));
+      //console.log('-----');
+      //console.log(messageData);
+      return Q.reject(messageData);
+    } else {
+      return Q.reject(err);
+    }
+  });
+};
+
+var extractDataFromEmailMessage = function(emailMessage, mapping) {
```

Comment:
Potential future optimization: might be able to use lodash `_.get()` here which lets you retrieve a property value for a given string path on on object (https://lodash.com/docs#get)